### PR TITLE
HIP-584: Copy precompile pricing logic without gasFeeInTinybars - Part 1

### LINKED ISSUE #5666: HIP-584: Copy precompile pricing logic without `gasFeeInTinybars` into hedera-mirror-node
### Problem

We need to extract the needed logic for calculating the gas cost of each precompile.

### Solution

Copy `PrecompilePricingUtils` and its associated logic without `gasFeeInTinybars` from `hedera-services` to `hedera-mirror-node`. `gasFeeInTinybars` is needed only for `TokenCreatePrecompile` to calculate the hbars we charge in addition to the gas fee. This issue aims to cover the main precompile pricing logic that is common for most precompiles. 

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by 0xivanov**
**Description**:

We need to extract the needed logic for calculating the gas cost of each precompile.

**Related issue(s)**:

Fixes #5666 

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

**Review by steven-sheehy — DISMISSED** [review]

File: hedera-mirror-web3/src/main/java/com/hedera/services/hapi/fees/usage/TxnUsage.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import static com.hedera.services.hapi.utils.fees.FeeBuilder.getAccountKeyStorageSize;
+
+import com.hederahashgraph.api.proto.java.Key;
+import com.hederahashgraph.api.proto.java.TransactionBody;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+public abstract class TxnUsage {
+    protected static final int AMOUNT_REPR_BYTES = 8;
+    protected static final UsageProperties usageProperties = USAGE_PROPERTIES;
```
Comment (david-bakin-sl) [inline hedera-mirror-web3/src/main/java/com/hedera/services/hapi/fees/usage/TxnUsage.java:30]: Using an enum class for a singleton utility class (_no state_) is a technique I haven't seen elsewhere, and it's very nice IMO.  (As long as you ensure the enum instances don't have state, which is done here.)  So why ruin the beauty of it by assigning it to a local?  It's a singleton utility with no state: just use it directly!  `USAGE_PROPERTIES.legacyReceiptStorageSpecs()`  (Same for `ESTIMATOR_UTILS`.)  

(Unless it is possible that in the future you might want a different implementation of the interface.  But if _that's_ the case then the name `SingletonUsageProperties` is misleading and might be better as `DefaultUsageProperties` or something.  And it's also not the way you're using it in this assignment right here.)

**Review by IvanKavaldzhiev — APPROVED** [review]
LGTM

File: hedera-mirror-web3/src/main/java/com/hedera/services/hapi/fees/usage/SingletonEstimatorUtils.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+import com.hederahashgraph.api.proto.java.SubType;
+import com.hederahashgraph.api.proto.java.TransactionBody;
+import com.hederahashgraph.api.proto.java.TransferList;
+
+@SuppressWarnings("java:S6548")
+public enum SingletonEstimatorUtils implements EstimatorUtils {
+    ESTIMATOR_UTILS;
+
+    @Override
+    public long baseNetworkRbs() {
```
Comment (david-bakin-sl) [inline hedera-mirror-web3/src/main/java/com/hedera/services/hapi/fees/usage/SingletonEstimatorUtils.java:37]: What does `Rbs` mean?  Evidently the `bs` is bytes/sec.  Leaving only the `R`.  Which is "receipt".  But I don't get why this is a "receipts bytes/sec".  It seems maybe more like a unit of storage use?

File: hedera-mirror-web3/src/main/java/com/hedera/services/hapi/fees/usage/EstimatorUtils.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+public interface EstimatorUtils {
+
+    /* A century of seconds */
+    long MAX_ENTITY_LIFETIME = 100L * 365L * 24L * 60L * 60L;
+
+    default long baseBodyBytes(TransactionBody txn) {
+        return (long) BASIC_TX_BODY_SIZE + txn.getMemoBytes().size();
+    }
+
+    default long nonDegenerateDiv(long dividend, int divisor) {
```
Comment (david-bakin-sl) [inline hedera-mirror-web3/src/main/java/com/hedera/services/hapi/fees/usage/EstimatorUtils.java:35]: There's no other place for this?  (As it is not specific to `EstimatorUtils`.)

File: hedera-mirror-web3/src/main/java/com/hedera/services/hapi/fees/usage/SingletonEstimatorUtils.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        var usages = FeeData.newBuilder();
+        var node = FeeComponents.newBuilder()
+                .setConstant(FEE_MATRICES_CONST)
+                .setBpt(usage.getBpt())
+                .setBpr(usage.getBpr())
+                .setSbpr(usage.getSbpr());
+        return usages.setNodedata(node).build();
+    }
+
+    int baseRecordBytes(TransactionBody txn) {
```
Comment (david-bakin-sl) [inline hedera-mirror-web3/src/main/java/com/hedera/services/hapi/fees/usage/SingletonEstimatorUtils.java:89]: Or could be a default method of the interface?

File: hedera-mirror-web3/src/main/java/com/hedera/services/hapi/fees/usage/EstimatorUtils.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    default long changeInBsUsage(long oldB, long oldLifetimeSecs, long newB, long newLifetimeSecs) {
+        oldLifetimeSecs = Math.min(MAX_ENTITY_LIFETIME, oldLifetimeSecs);
+        newLifetimeSecs = Math.min(MAX_ENTITY_LIFETIME, newLifetimeSecs);
+
+        newLifetimeSecs = Math.max(oldLifetimeSecs, newLifetimeSecs);
+        long oldBs = oldB * oldLifetimeSecs;
+        long newBs = newB * newLifetimeSecs;
+        return Math.max(0, newBs - oldBs);
```
Comment (david-bakin-sl) [inline hedera-mirror-web3/src/main/java/com/hedera/services/hapi/fees/usage/EstimatorUtils.java:55]: So I guess only interested in _upward_ changes?  Should this be reflected in the method name?  (`increaseinBsusage`?)

File: hedera-mirror-web3/src/main/java/com/hedera/services/hapi/fees/usage/SingletonUsageProperties.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    USAGE_PROPERTIES;
+
+    @Override
+    public int accountAmountBytes() {
+        return LONG_SIZE + BASIC_ENTITY_ID_SIZE;
+    }
+
+    @Override
+    public int nftTransferBytes() {
+        return LONG_SIZE + 2 * BASIC_ENTITY_ID_SIZE;
```
Comment (david-bakin-sl) [inline hedera-mirror-web3/src/main/java/com/hedera/services/hapi/fees/usage/SingletonUsageProperties.java:33]: Here (and above in `accountAmountBytes` perhaps a comment to explain what these two "long"s you're accounting for are.  At least a reference to the class or whatever it is that this is providing the size-in-bytes of.

File: hedera-mirror-web3/src/main/java/com/hedera/services/hapi/fees/usage/TxnUsage.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    protected TxnUsage(final TransactionBody op, final TxnUsageEstimator usageEstimator) {
+        this.op = op;
+        this.usageEstimator = usageEstimator;
+    }
+
+    public static <T> long keySizeIfPresent(final T op, final Predicate<T> check, final Function<T, Key> getter) {
+        return check.test(op) ? getAccountKeyStorageSize(getter.apply(op)) : 0L;
+    }
+
+    protected void addAmountBpt() {
```
Comment (david-bakin-sl) [inline hedera-mirror-web3/src/main/java/com/hedera/services/hapi/fees/usage/TxnUsage.java:44]: Is there any value in actual use of making these methods fluid? (e.g., returning `this` `TxnUsage`) (I see that `FeeComponents` supports that via a builder.  So maybe the pattern used should be builder.)

File: hedera-mirror-web3/src/main/java/com/hedera/services/hapi/fees/usage/TxnUsage.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+    protected final TransactionBody op;
+    protected final TxnUsageEstimator usageEstimator;
+
+    protected TxnUsage(final TransactionBody op, final TxnUsageEstimator usageEstimator) {
+        this.op = op;
+        this.usageEstimator = usageEstimator;
+    }
+
+    public static <T> long keySizeIfPresent(final T op, final Predicate<T> check, final Function<T, Key> getter) {
```
Comment (david-bakin-sl) [inline hedera-mirror-web3/src/main/java/com/hedera/services/hapi/fees/usage/TxnUsage.java:40]: This is so generic even in its _argument names_ it's hard to know what to instantiate it with.  You pass an "op" to a Predicate and do something with it?  Needs a comment at the least.

File: hedera-mirror-web3/src/main/java/com/hedera/services/hapi/fees/usage/TxnUsage.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        return check.test(op) ? getAccountKeyStorageSize(getter.apply(op)) : 0L;
```
Comment (david-bakin-sl) [inline hedera-mirror-web3/src/main/java/com/hedera/services/hapi/fees/usage/TxnUsage.java:41]: `getAccountKeyStorageSize` returns an `int` so why does this return a `long`?

File: hedera-mirror-web3/src/main/java/com/hedera/services/hapi/fees/usage/UsageEstimate.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    private long rbs;
+    private long sbs;
+
+    private final FeeComponents.Builder base;
+
+    public UsageEstimate(final FeeComponents.Builder base) {
+        this.base = base;
+    }
+
+    public void addRbs(final long amount) {
```
Comment (david-bakin-sl) [inline hedera-mirror-web3/src/main/java/com/hedera/services/hapi/fees/usage/UsageEstimate.java:36]: Perhaps `Rbs` and `Sbs` are things that are just known in the mirror node and I'm ignorant of them.  NVM then my previous comment. (But maybe you can point me to where they're defined?)

**Review by david-bakin-sl — COMMENTED** [review]
Oh fudge.  NVM.  I forgot that this was a "copy" and not new code.  Maybe the PR description should say "copied from services" (and even give a commit hash).

File: hedera-mirror-web3/src/main/java/com/hedera/services/hapi/fees/usage/UsageEstimate.java (hunk 10 above)
Comment (steven-sheehy) [reply hedera-mirror-web3/src/main/java/com/hedera/services/hapi/fees/usage/UsageEstimate.java:36]: I've no idea what they are either. They're consensus node concepts. I think the BS part is byte size/storage for estimating transaction costs.

**Review by steven-sheehy — APPROVED** [review]

**Pull request merged**